File: uicorek/src/main/java/com/mozhimen/uicorek/bannerk/helpers/BannerKViewPager.java
package com.mozhimen.uicorek.bannerk.helpers;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewpager.widget.ViewPager;

import java.lang.reflect.Field;

/**
 * 实现自动翻页的ViewPager
 */
public class BannerKViewPager extends ViewPager {

    private int mIntervalTime;//滚动时间间隔

    /**
     * 是否开启自动轮播
     */
    private boolean mAutoPlay = true;
    private boolean isLayout;
    private Handler mHandler = new Handler();
    private Runnable mRunnable = new Runnable() {
        @Override
        public void run() {
            //切换到下一个
            next();
            mHandler.postDelayed(this, mIntervalTime);
        }
    };

    /**
     * 是否开启自动轮播
     *
     * @param context
     */
    public BannerKViewPager(@NonNull Context context) {
        super(context);
    }

    public BannerKViewPager(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    /**
     * 设置ViewPagerK的滚动速度
     *
     * @param duration
     */
    public void setScrollerDuration(int duration) {
        try {
            Field scrollerField = ViewPager.class.getDeclaredField("mScroller");
            scrollerField.setAccessible(true);
            scrollerField.set(this, new BannerKScroller(getContext(), duration));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void setIntervalTime(int intervalTime) {
        this.mIntervalTime = intervalTime;
    }

    public void setAutoPlay(boolean autoPlay) {
        this.mAutoPlay = autoPlay;
        if (!mAutoPlay) {
            mHandler.removeCallbacks(mRunnable);
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        switch (ev.getAction()) {
            case MotionEvent.ACTION_UP:
                break;
            case MotionEvent.ACTION_CANCEL:
                start();
                break;
            default:
                stop();
                break;
        }
        return super.onTouchEvent(ev);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        super.onLayout(changed, l, t, r, b);
        isLayout = true;
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isLayout && getAdapter() != null && getAdapter().getCount() > 0) {
            try {
                //fix 使用RecyclerView + ViewPager bug
                Field mScroller = ViewPager.class.getDeclaredField("mFirstLayout");
                mScroller.setAccessible(true);
                mScroller.set(this, false);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        start();
    }

    @Override
    protected void onDetachedFromWindow() {
        //fix 使用RecyclerView + ViewPager 卡帧bug
        if (((Activity) getContext()).isFinishing()) {
            super.onDetachedFromWindow();
        }
        stop();
    }

    public void start() {
        mHandler.removeCallbacksAndMessages(null);
        if (mAutoPlay) {
            mHandler.postDelayed(mRunnable, mIntervalTime);
        }
    }

    public void stop() {
        mHandler.removeCallbacksAndMessages(null);
    }

    /**
     * 设置下一个要显示的item,并返回item的pos
     *
     * @return 下一个要显示item的pos
     */
    private int next() {
        int nextPosition = -1;
        if (getAdapter() == null || getAdapter().getCount() <= 1) {
            //TODO: stop()
            return nextPosition;
        }
        nextPosition = getCurrentItem() + 1;
        //下一个索引大于adapter的view的最大数量时重新开始
        if (nextPosition >= getAdapter().getCount()) {
            //获取第一个item的索引
            nextPosition = ((BannerKAdapter) getAdapter()).getFirstItem();
        }
        setCurrentItem(nextPosition, true);
        return nextPosition;
    }
}

File: uicorek/src/main/java/com/mozhimen/uicorek/bannerk/helpers/BannerKViewPager.kt
package com.mozhimen.uicorek.bannerk.helpers

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.viewpager.widget.ViewPager
import java.lang.Exception

/**
 * @ClassName BannerKViewPager
 * @Description 实现了自动翻页的ViewPager
 * @Author mozhimen / Kolin Zhao
 * @Date 2022/4/19 17:40
 * @Version 1.0
 */
class BannerKViewPager @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
    ViewPager(context, attrs) {

    private var _intervalTime = 0
    private var _autoPlay = true//是否开启自动轮播
    private var _isLayout = false
    private val _handler = Handler(Looper.getMainLooper())
    private val _runnable: Runnable = object : Runnable {
        override fun run() {
            nextItem()
            _handler.postDelayed(this, _intervalTime.toLong()) //延时一定时间执行下一次
        }
    }

    /**
     * 是否自动轮播
     * @param autoPlay Boolean
     */
    fun setAutoPlay(autoPlay: Boolean) {
        this._autoPlay = autoPlay
        if (!_autoPlay) {
            _handler.removeCallbacks(_runnable)
        }
    }

    /**
     * 设置ViewPager的滚动速度
     * @param duration Int page切换的时间长度
     */
    fun setScrollDuration(duration: Int) {
        try {
            val scrollerField = ViewPager::class.java.getDeclaredField("mScroller")
            scrollerField.isAccessible = true
            scrollerField[this] = BannerKScroller(context, duration)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 设置页面停留时间
     * @param intervalTime Int 停留时间单位毫秒
     */
    fun setIntervalTime(intervalTime: Int) {
        this._intervalTime = intervalTime
    }

    /**
     * 开始
     */
    fun start() {
        _handler.removeCallbacksAndMessages(null)
        if (_autoPlay) {
            _handler.postDelayed(_runnable, _intervalTime.toLong())
        }
    }

    /**
     * 结束
     */
    fun stop() {
        _handler.removeCallbacksAndMessages(null) //停止Timer
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(ev: MotionEvent): Boolean {
        when (ev.action) {
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> start()
            else -> stop()
        }
        return super.onTouchEvent(ev)
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        _isLayout = true
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (_isLayout && adapter != null && adapter!!.count > 0) {
            try {
                //fix 使用RecyclerView + ViewPager bug https://blog.csdn.net/u011002668/article/details/72884893
                val mScroller = ViewPager::class.java.getDeclaredField("mFirstLayout")
                mScroller.isAccessible = true
                mScroller[this] = false
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        start()
    }

    override fun onDetachedFromWindow() {
        //fix 使用RecyclerView + ViewPager bug
        if ((context as Activity).isFinishing) {
            super.onDetachedFromWindow()
        }
        stop()
    }

    private fun nextItem(): Int {
        var nextPosition = -1
        if (adapter == null || adapter!!.count <= 1) {
            stop()
            return nextPosition
        }
        nextPosition = currentItem + 1
        //下一个索引大于adapter的view的最大数量时重新开始
        if (nextPosition >= adapter!!.count) {
            nextPosition = (adapter as BannerKAdapter).getFirstItem()
        }
        setCurrentItem(nextPosition, true)
        return nextPosition
    }
}